File: A-Dex/discord-bot/src/signature.js
const crypto = require('crypto');

// Shared HMAC format with backend: sha256(secret, `${timestamp}.${rawBody}`).
function signBody(secret, timestamp, body) {
  const payload = typeof body === 'string' ? body : JSON.stringify(body || {});
  return crypto.createHmac('sha256', secret).update(`${timestamp}.${payload}`).digest('hex');
}

module.exports = { signBody };

File: A-Dex/discord-bot/bot/signature.py
import hashlib
import hmac
import json
from typing import Any


def sign_body(secret: str, timestamp: str, body: str | dict[str, Any] | None) -> str:
    """Match backend HMAC format: sha256(secret, f"{timestamp}.{raw_json_body}")."""
    if isinstance(body, str):
        payload = body
    else:
        payload = json.dumps(body or {}, separators=(",", ":"), ensure_ascii=False)

    return hmac.new(secret.encode("utf-8"), f"{timestamp}.{payload}".encode("utf-8"), hashlib.sha256).hexdigest()
